File: objc/Project7/Project7/ViewController.m
//
//  ViewController.m
//  Project7
//
//  Created by Jinwoo Kim on 2/11/21.
//

#import "ViewController.h"

@implementation ViewController

- (void)setup {
    self.pages = @{};
    self.firstRun = YES;
}

- (void)viewDidLoad {
    [self setup];
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.mapView.centerCoordinate = CLLocationCoordinate2DMake(51.38, -2.36);
    [self updateMap];
}

- (void)focusOn:(City *)city {
    self.mapView.centerCoordinate = city.coordinates;
}

- (void)updateMap {
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"https://en.wikipedia.org/w/api.php?ggscoord=%f%%7C%f&action=query&prop=coordinates%%7Cpageimages%%7Cpageterms&colimit=50&piprop=thumbnail&pithumbsize=500&pilimit=50&wbptterms=description&generator=geosearch&ggsradius=10000&ggslimit=50&format=json", self.mapView.centerCoordinate.latitude, self.mapView.centerCoordinate.longitude]];
    if (url == nil) return;
    
    dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INTERACTIVE, 0), ^{
        NSData *data = [NSData dataWithContentsOfURL:url];
        self.pages = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil][@"query"][@"pages"];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self reloadAnnotations];
        });
    });
}

- (void)reloadAnnotations {
    // remove all existing annotations
    [self.mapView removeAnnotations:self.mapView.annotations];
    NSLog(@"%@", self.pages);
    
    // create a new, empty map rect
    MKMapRect visibleMap = MKMapRectNull;
    
    // loop over all the pages we received from Wikipedis
    for (NSDictionary *page in self.pages.allValues) {
        // create a point annotation for it, configuring its title and coordinate
        MKPointAnnotation *point = [MKPointAnnotation new];
        point.title = (NSString *)page[@"title"];
        point.coordinate = CLLocationCoordinate2DMake([(NSNumber *)page[@"coordinates"][0][@"lat"] doubleValue],
                                                      [(NSNumber *)page[@"coordinates"][0][@"lon"] doubleValue]);
        
        // add that to the map
        [self.mapView addAnnotation:point];
        
        // add its position to our overall map rect+
        // https://gist.github.com/andrewgleave/915374
        MKMapPoint mapPoint = MKMapPointForCoordinate(point.coordinate);
        MKMapRect pointRect = MKMapRectMake(mapPoint.x, mapPoint.y, 0.1, 0.1);
        visibleMap = MKMapRectUnion(visibleMap, pointRect);
    }
    
    // if this is the first time we have loaded pins
    if (self.firstRun) {
        // zoom to the configure map rect
        [self.mapView setVisibleMapRect:visibleMap
                            edgePadding:UIEdgeInsetsMake(60, 60, 90, 90)
                               animated:YES];
        
        // mark that we've done it at lease once
        self.firstRun = NO;
    }
}

- (void)mapView:(MKMapView *)mapView regionDidChangeAnimated:(BOOL)animated {
    [self updateMap];
}

@end

File: swift/Project7/Project7/ViewController.swift
//
//  ViewController.swift
//  Project7
//
//  Created by Jinwoo Kim on 2/11/21.
//

import UIKit
import MapKit

class ViewController: UIViewController, MKMapViewDelegate {
    @IBOutlet weak var mapView: MKMapView!
    var pages = [String: JSON]()
    var firstRun = true
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        mapView.centerCoordinate = CLLocationCoordinate2D(latitude: 51.38, longitude: -2.36)
        updateMap()
    }

    func focus(on city: City) {
        mapView.centerCoordinate = city.coordinates
    }
    
    func updateMap() {
        guard let url = URL(string: "https://en.wikipedia.org/w/api.php?ggscoord=\(self.mapView.centerCoordinate.latitude)%7C\(self.mapView.centerCoordinate.longitude)&action=query&prop=coordinates%7Cpageimages%7Cpageterms&colimit=50&piprop=thumbnail&pithumbsize=500&pilimit=50&wbptterms=description&generator=geosearch&ggsradius=10000&ggslimit=50&format=json") else {
            return
        }
        
        DispatchQueue.global(qos: .userInteractive).async {
            if let data = try? Data(contentsOf: url) {
                let json = JSON(data)
                self.pages = json["query"]["pages"].dictionaryValue
                DispatchQueue.main.async {
                    self.reloadAnnotations()
                }
            }
        }
    }
    
    func reloadAnnotations() {
        // remove all existing annotations
        mapView.removeAnnotations(mapView.annotations)
        
        // create a new, empty map rect
        var visibleMap = MKMapRect.null
        
        // loop over all the pages we received from Wikipedia
        for page in pages.values {
            // create a point annotation for it, configuring its title and coordinate
            let point = MKPointAnnotation()
            point.title = page["title"].stringValue
            point.coordinate = CLLocationCoordinate2D(latitude: page["coordinates"][0]["lat"].doubleValue,
                                                      longitude: page["coordinates"][0]["lon"].doubleValue)
            
            // add that to the map
            mapView.addAnnotation(point)
            
            // add its position to our overall map rect+
            let mapPoint = MKMapPoint(point.coordinate)
            let pointRect = MKMapRect(x: mapPoint.x, y: mapPoint.y, width: 0.1, height: 0.1)
            visibleMap = visibleMap.union(pointRect)
            
            print("Noti!!! \(point.coordinate)")
        }
        
        // if this is the first time we have loaded pins
        if firstRun {
            // zoom to the configured map rect
            mapView.setVisibleMapRect(visibleMap, edgePadding: UIEdgeInsets(top: 60, left: 60, bottom: 90, right: 90), animated: true)
            
            // mark that we've done it at least once
            firstRun = false
        }
    }
    
    func mapView(_ mapView: MKMapView, regionDidChangeAnimated animated: Bool) {
        updateMap()
    }
}
